### PR TITLE
6688 add debianization support v1

### PR DESCRIPTION
**Pull request by tttech-industrial-buchsbaum**
This patchset brings in debianization for ACRN.

The first set of patches

  commit 25245ddf "Update gitignore to ignore Eclipse project config"
  commit 020902ec "board-inspector: Fix sys.path for all internal modules"
  commit 83b395d1 "dm: Remove dependency on libuuid"
  commit 7119f60e "Use libdrm library for drm access"
  commit c66dfe61 "hv: Fixup serial.conf permissions"

only relate to some dependency cleanup (since packaging requires clean dependencies) and minor fixes, e.g for board-inspector to be able to run it as a package (i.e. not within the acrn-hypervisor source tree) cleanly.

The main commit is

  commit 886ee330 "Add debianization support"

which brings in all the required stuff to create Debian packages.

Due to the recent change

  commit c25de24a "config-tools: remove board and scenario attributes"

I had to adapt the scenario file detection logic in step

  commit 4ef5fea3 "d/rules: Update regarding config detection".

To accomodate to the needs of the config-tools, I had to pin python3-elementpath and python3-xmlschema to the required versions in

  commit 83cff553 "d/control: Pin elementpath and xmlschema".

Since building ACRN with all its tools and helpers requires at least some exact or more recent version than usually available 
within various distributions, a docker based package build system is introduced in

  commit e49ca64d "debian: Add docker build helpers"

to take the burden of setting up a build system on your development host.

Furthermore, the build results are enriched by the special versions of packages also needed at runtime to form a simple, local APT repository for easy install on the target (see below):

  commit 46651c26 "debian: Create local apt repo after build"

The debian/README.md file containing additional information has then been updated accordingly in

  commit 9ecca8df "debian: Update and rename README.md"

I also added the configurations I used for testing

  commit c5016f66 "debian: Add nuc7i7dnh/shared+initrd"
  commit b9ac69ae "debian: Add kontron-COMe-mAL10/shared+initrd"

which also provide examples for adding configurations within the `debian/configs` folder and - last but not least - a scenario using initrd, which is a essential requirement for Debian based distributions.

The last

  commit 72c8875e "debian: Update changelog for UNRELEASED package build"

is just an example for debian/changelog maintenance in a still unreleased state.

The packages have been tested (including starting an Alpine Linux VM) on Debian Buster, Debian Bullseye and Ubuntu Server 20.04 on kontron-COMe-mAL10 and on Debian Buster on nuc7i7dnh, following the procedure below (here for Debian Buster):

* Build the ACRN packages:
```
  user@host:~/work/acrn-hypervisor$ export DISTRO=buster; rm -rf build/${DISTRO}; VENDOR=debian debian/docker/acrn-docker-build.sh --git-ignore-branch
```
* Install Debian Buster via netinstall ISO on your target device,  just install a minimum system, but add the SSH server.

* Configure the ssh root access for easy remote access to the target system.
 
* Manually install an appropriate, ACRN enabled Linux kernel (>5.10). I used a kernel merged from buster-backports, and the intel-lts/acrn-kernel repository (already packaged):
```  
  user@host:~/work/linux$ scp linux-image-5.10.0-0.bpo.9.1-acrn-amd64-unsigned_5.10.70-1_bpo10+1+intel+lts1_amd64.deb root@target:/tmp
  root@target:~# apt-get install -y /tmp/linux-image-5.10.0-0.bpo.9.1-acrn-amd64-unsigned_5.10.70-1_bpo10+1+intel+lts1_amd64.deb
```
  This fulfills the configuration and change requirements of a Debian system as well as supporting ACRN.
  
  Remark: a kernel built directly from acrn-kernel repository using kernel_config_service_vm as proposed in the GSG **DOES NOT** boot on a Debian system! I assume some issues with the kernel configuration regarding initrd, which is an essential requirement for Debian based distributions, but I have not further investigated this.
  
* Copy over the build result folder to the target device, e.g.
```
  root@target:~# mkdir -p /opt/local-apt
  user@host:~/work/acrn-hypervisor$ scp build/buster/* root@target:/opt/local-apt
```
* Create an APT sources.list for the local APT repository and update the APT database
```
  root@target:~# echo "deb [trusted=yes] file:/opt/local-apt ./"  > /etc/apt/sources.list.d/local-apt.list
  root@target:~# apt-get update
```
* Now install ACRN via the metapackage acrn-system. Choose the corresponding board and scenario! Since grub-acrn and the integration into the Debian system need an initrd, use the share+initrd scenario (available for nuc7i7dnh and kontron-COMe-mAL10 only).
```
  root@target:~# apt-get install -y acrn-system
```
* Reboot the system:
```
  root@target:~# reboot
```
* After reboot, verify you are running the service VM and ACRN has bee detected:
```
  root@target:~# dmesg | grep ACRN
  [    0.000000] Hypervisor detected: ACRN
 ```
 * With this kernel ACRN module is built as a kernel module, so load it first, to verify HSM:
 ```
   root@target:~# modprobe acrn
   root@target:~# ls -al /dev/acrn_hsm
   crw------- 1 root root 10, 61 Apr 29 17:32 /dev/acrn_hsm
```
Now we are ready to launch VMs on our ACRN enhanced Debian system ;-)

### CODE REVIEW

**Comment by acrnsi** [comment]
Can one of the admins verify this patch?

File: debian/control [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+   flex,
+   bison,
+   xsltproc,
+   clang-format,
+   bc,
+   acpica-tools (>= 20200326),
+   python3-lxml,
+   python3-defusedxml,
+   python3-xmlschema (= 1.9.2-1),
+   python3-elementpath (= 2.4.0-1)
```
Comment (Pirouf) [inline debian/control:37]: (>= 2.4.0-1) Debian maintainer is on jump straight to https://salsa.debian.org/debian/elementpath/-/tree/debian/2.5.0-1
 no 2.4.0-1 on debian repos

File: debian/control [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+   libcjson-dev,
+   flex,
+   bison,
+   xsltproc,
+   clang-format,
+   bc,
+   acpica-tools (>= 20200326),
+   python3-lxml,
+   python3-defusedxml,
+   python3-xmlschema (= 1.9.2-1),
```
Comment (Pirouf) [inline debian/control:36]: (>= 1.9.2-1) Debian maintainer released jump straight to https://salsa.debian.org/python-team/packages/python-xmlschema/-/tree/debian/1.10.0-1
no 1.9.2-1

**Review by Pirouf — COMMENTED** [review]
@tttech-industrial-buchsbaum Really Great work!
 
It is a very positive & healthy step to Debian sw ecosystem compat : Canonical/Ubuntu, GNU/Debian, ...
`dpgk-buildpackage` offers a standard and easy-way to streamline XML user-config injection into target Industrial OT/IT systems requiring acrn-hypervisor/device-model rebuild. 
Coupled with auto-generated `grub-acrn` config definitely lowers down obstacles for user to get to iterate acrn configs over Intel SoC and UEFI FW platform tunnables.

A simple suggestion about `deb-src` support which would facilitate arcn downstream ODM/OEM product sw-lifecycle.

**Comment by tttech-industrial-buchsbaum** [comment]
I like the idea about stressing the possibility of apt-get source directly on the target system! I'll add a respective section in the README file!

ad version pinning: I know about the maintainer's jump to 2.5.0-1 and 1.10.0-1, respectively. Unfortunately the versions used here are strong requirement and build fails (at least the version I'm building with) when using the newest version, see also  https://github.com/projectacrn/acrn-hypervisor/issues/7372. There is work already ongoing to eliminate this issue. Until we have solved it, we will have to stick to the versions we have in debian/control right now and go the painful way of repackaging and providing the respective package versions. As soon as the fix is available, it is an easy step to move to python3-xmlschema (>= 1.10.0-1) and python3-elementpath (>= 2.5.0-1).

________________________________________

**Comment by szhen11** [comment]
start to run premerge test after installing libdrm-dev

**Comment by szhen11** [comment]
> start to run premerge test after installing libdrm-dev

ACRN build Passed after `sudo apt install libdrm-dev`.

**Comment by junjiemao1** [comment]
@tttech-industrial-buchsbaum The workaround when using the latest elementpath has been merged (see #7390). Do you see any other issue here?

**Review by NanlinXie — APPROVED** [review]
Thanks @tttech-industrial-buchsbaum for your contribution.

**Review by lifeix — APPROVED** [review]

**Review by ywan170 — COMMENTED** [review]
dm patch looks good to me.

**Review by ywan170 — APPROVED** [review]

File: debian/configs/kontron-COMe-mAL10/kontron-COMe-mAL10.xml [hunk 3]
```diff
@@ -0,0 +1,997 @@
+<acrn-config board="kontron-COMe-mAL10">
```
Comment (junjiemao1) [inline debian/configs/kontron-COMe-mAL10/kontron-COMe-mAL10.xml:1]: Shall we exclude the user-specific XML files?

**Review by junjiemao1 — CHANGES_REQUESTED** [review]

**Comment by tttech-industrial-buchsbaum** [comment]
Please see my follow-up PR https://github.com/projectacrn/acrn-hypervisor/pull/7438

**Pull request closed without merging**